### PR TITLE
Remove public gtest dependency from libcudf conda package

### LINKED ISSUE #13381: [FEA] Remove libcudf dependency on gtest / gmock
**Is your feature request related to a problem? Please describe.**

@ajschmidt8, myself, and others have been confused in the past by `libcudf`'s conda dependency on `gtest` and `gmock`. This is currently needed because the `cudftestutil` target (which depends on these) is shipped as part of the `libcudf` package. We'd like to remove the dependency of `libcudf` on `gtest`.

**Describe the solution you'd like**

@vyasr thinks we can split `cudftestutil` and `cudftest_default_stream` into a separate conda package. Perhaps it can be named `libcudf-test-utils` or similar.

### PR DESCRIPTION
**Pull request by robertmaynard**
## Description

Reworks the cudftestutil and dependency chain to remove the public gtest dependency in libcudf conda package.
The libcudftestutil was previously made static due to issues using a static system GTest that wasn't build with `fPIC`. Using  a GTest from `rapids-cmake` which is built with `fPIC` enabled, removes this restriction and allows us to remove the public depedency.

Some notes:
-  We need to align all of RAPIDS C++ projects on static GTest from `rapids-cmake` 
- None of the compiled components / classes of `libcudftestutils` publically depend on GTest
- Two of the libcudftestutils header only components bring include gtest. Since these headers aren't required to be used we are going to consider them optional.
- Therefore using these optional `libcudftestutils` header will require downstream users to bring in GTest.

Fixes #13381

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

**Review by bdice — APPROVED** [review]

**Comment by bdice** [comment]
I merged the upstream to get CI unblocked.

**Review by vyasr — CHANGES_REQUESTED** [review]
While I think this change is already safe across the rest of RAPIDS (https://github.com/rapidsai/build-planning/issues/32), I suspect we need to do a bit more work in cudf. Specifically, I think we have to address #13381. Currently libcudftestutil.a exports a public dependency on GTest. Granting that it is already a static library, that still means that consumers of it will see APIs that have use gtest. What happens if someone tries to link against libcudftestutil.a, but has compiled their code using a different version of gtest? Won't that result in incompatible binaries?

In practice the only use case I'm aware of is cuspatial, which uses rapids-cmake and we control, so it's probably not a serious issue, but unless I'm misunderstanding it is a real potential issue that is worth trying to resolve before merging this.

**Comment by robertmaynard** [comment]
> While I think this change is already safe across the rest of RAPIDS ([rapidsai/build-planning#32](https://github.com/rapidsai/build-planning/issues/32)), I suspect we need to do a bit more work in cudf. Specifically, I think we have to address #13381. Currently libcudftestutil.a exports a public dependency on GTest. Granting that it is already a static library, that still means that consumers of it will see APIs that have use gtest. What happens if someone tries to link against libcudftestutil.a, but has compiled their code using a different version of gtest? Won't that result in incompatible binaries?
> 
> In practice the only use case I'm aware of is cuspatial, which uses rapids-cmake and we control, so it's probably not a serious issue, but unless I'm misunderstanding it is a real potential issue that is worth trying to resolve before merging this.

This is a great point. Using a static GTest with libcudf would mean that conda consumers would be forced to never have GTest in the conda env. I agree with you that this is not a desired outcome, and it looks like we can't move over to a static GTest for libcudf without first splitting the testing component into a separate conda package.

I am going to close this PR in that case, as that is out of scope of this change

**Review by bdice — APPROVED** [review]

**Comment by robertmaynard** [comment]
Re-opnening this PR since we realized that by always building our own GTest, we don't need to worry about linking to a static system GTest which wasn't built with fPIC.

**Review by bdice — APPROVED** [review]
Re-reviewed on request. Looks good.

File: cpp/tests/utilities/random_seed.cpp
Comment (vyasr) [inline cpp/tests/utilities/random_seed.cpp:1]: :upside_down_face:

**Review by vyasr — APPROVED** [review]

File: cpp/include/cudf_test/cudf_gtest.hpp [hunk 1]
```diff
@@ -16,86 +16,6 @@
 
 #pragma once
 
-#ifdef GTEST_INCLUDE_GTEST_GTEST_H_
```
Comment (robertmaynard) [inline cpp/include/cudf_test/cudf_gtest.hpp:19]: This is not needed now that we use a GTest version dowloaded via rapids-cmake. In addition the logic was incorrectly triggering on the devcontainer builds due to a system installed GTest causing us to break the internals of GTest

**Pull request merged**